src/wp-includes/post-template.php

```diff
@@ -1798,7 +1798,12 @@ function get_the_password_form( $post = 0 ) {
 		$class                 = ' password-form-error';
 	}
 
-	$output = '<form action="' . esc_url( site_url( 'wp-login.php?action=postpass', 'login_post' ) ) . '" class="post-password-form' . $class . '" method="post">' . $invalid_password_html . '
+	$redirect_field = sprintf(
+		'<input type="hidden" name="redirect_to" value="%s" />',
+		esc_attr( get_permalink( $post ) )
+	);
```

Comment:
Similar to `invalid_password_html` check if we have post ID if not return empty string.

```
$redirect_to = '';
if ( ! empty( $post->ID )  ) {
	$redirect_field = sprintf(
		'<input type="hidden" name="redirect_to" value="%s" />',
		esc_attr( get_permalink( $post->ID ) )
	);
}
```